installer: simple PXE bootable NixOS installer

nshalman opened the pull request:
###### Things done
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Resolves #14538

nshalman committed:
installer: simple PXE bootable NixOS installer

The Nix store squashfs is stored inside the initrd instead of separately

bobvanderlinden committed:
nixos: tests: added netboot test

cstrahan commented:
Wow, this looks pretty awesome! I don't have enough bandwidth at the moment to give this a proper review (I'd have to bone up on the intersection of PXE and qemu first), but this looks incredible :).

If you don't mind me asking, do you have a particular use case in mind for this? I see you work for OmniTI and are involved in some neat looking stuff (e.g. Cerana), so I figure you might be up to something pretty cool :).

nshalman commented:
@cstrahan yes, as we rebrand from from our old project name over to Cerana, we're also looking at replacing our old build system with using Nix/NixOS/Nixpkgs as the substrate. Cerana's model is to boot most of the cluster over the network and the build system we'd be replacing currently generates a kernel and ramdisk with no need for obtaining an additional squashfs.

Ultimately I'm trying to build a live environment rather than an installer that lives entirely in memory and consists of only a kernel and ramdisk.

There's a ton of customization to our old build system that would we get mostly for free from Nixpkgs if I get all the other pieces into place. I'm hopeful that various pieces of our work will in turn be of value to the broader Nix community.

cstrahan commented:
> Ultimately I'm trying to build a live environment rather than an installer that lives entirely in memory and consists of only a kernel and ramdisk.

Sounds lovely.

> There's a ton of customization to our old build system that would we get mostly for free from Nixpkgs if I get all the other pieces into place. I'm hopeful that various pieces of our work will in turn be of value to the broader Nix community.

Looking forward to seeing that unfold :).

bobvanderlinden commented on nixos/tests/boot.nix:
```diff
@@ -44,5 +44,41 @@ in {
+        text = ''
+          #!ipxe
+          dhcp
+          kernel bzImage init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams} console=ttyS0
```
I *think* it would be better to leave out `console=ttyS0` here, since it will not be useful on physical systems and might cause confusion in combination with `config.boot.kernelParams`.

It does mean the test script will not work, however with the automated netboot test Nix would add console anyway.

bobvanderlinden commented on nixos/release.nix:
```diff
@@ -104,6 +104,22 @@ in rec {
+      build.netbootRamdisk
+      build.kernel
+      build.netbootKernelAppend
+      build.netbootTestScript
```
Not sure if `netbootTestScript` should be removed. With the boot test it isn't as useful, but it might still be practical for some.

I do think `netbootKernelAppend` can be removed, as there is no use for that anymore that I can think of.

bobvanderlinden commented:
Overal this looks good. I still need to give this a try on physical hardware. Once my laptop is in proper condition again I'll do a test.

Has anyone else tested this on physical machine yet?

Also, it might be nice to eventually add an module for iPXE over dhcp with nixos installer too (probably not in this PR). Similar to what @ehmry mentioned [here](https://github.com/NixOS/nixpkgs/issues/2100#issuecomment-157096584).

nshalman committed:
installer: netboot: remove extraneous bits

remove netbootKernelAppend and netbootTestScript
(per feedback from @bobvanderlinden)

nshalman commented:
@bobvanderlinden If those removals look good to you I can squash them back down into the original commit and force-push the cleaned up branch.

bobvanderlinden commented:
I'm not a NixOS 'member', so others should decide whether it's worth merging in its current state.
@fpletz and @domenkozar seemed interested/experienced in iPXE booting for NixOS. What do you think? Is this viable for merging once it is tested on physical machines?

joachifm commented:
Looks good to me. Should be merged for sure, provided it works as advertised.

dezgeg commented on nixos/modules/installer/netboot/netboot.nix:
```diff
@@ -0,0 +1,97 @@
+    # Create the initrd
+    system.build.netbootRamdisk = pkgs.makeInitrd {
+      inherit (config.boot.initrd) compressor prepend;
+
```
This will cause a lack of some files (defined here: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/stage-1.nix#L237), which sounds like it could cause problems on some hardware.

I think the following might solve that by building the new initrd on top of the stock NixOS generated initrd (totally untested):

````
      inherit (config.boot.initrd) compressor;
      prepend = config.system.build.initialRamdisk;

      contents = [
          { object = config.system.build.squashfsStore;
            symlink = "/nix-store.squashfs";
          };
    ];
````

nshalman replied:
@dezgeg, The example code you gave fails (assuming I copied it correctly...).

I made this change:
```diff
diff --git a/nixos/modules/installer/netboot/netboot.nix b/nixos/modules/installer/netboot/netboot.nix
index 41f79eb..7080b52 100644
--- a/nixos/modules/installer/netboot/netboot.nix
+++ b/nixos/modules/installer/netboot/netboot.nix
@@ -62,13 +62,11 @@ with lib;

     # Create the initrd
     system.build.netbootRamdisk = pkgs.makeInitrd {
-      inherit (config.boot.initrd) compressor prepend;
+      inherit (config.boot.initrd) compressor;
+      prepend = config.system.build.initialRamdisk;

       contents =
-        [ { object = config.system.build.bootStage1;
-            symlink = "/init";
-          }
-          { object = config.system.build.squashfsStore;
+        [ { object = config.system.build.squashfsStore;
             symlink = "/nix-store.squashfs";
           }
         ];
```

And got this error:
```
these derivations will be built:
  /nix/store/7h771qyvq24b1iz7233vb6wfrynfnmlp-initrd.drv
  /nix/store/fdmxgaw0rcjis4az63jr24q9vc8vfgx3-netboot.drv
building path(s) ‘/nix/store/pn2rr4wksz3f790hmnx0cpmsf0rwmii5-initrd’
cat: /nix/store/fdl35wv4i960fmx8yrj9ia971wpv1nm2-initrd: Is a directory
builder for ‘/nix/store/7h771qyvq24b1iz7233vb6wfrynfnmlp-initrd.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/fdmxgaw0rcjis4az63jr24q9vc8vfgx3-netboot.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/fdmxgaw0rcjis4az63jr24q9vc8vfgx3-netboot.drv’ failed
```

Something needs a trailing ```/initrd``` to find the file inside the directory.

bobvanderlinden replied:
`system.build.initialRamdisk` results in a directory with `initrd` in it. It'll work if you're referring to /initrd:

https://github.com/bobvanderlinden/nixpkgs/commit/e6c82ede5c7f10d0b27ac293528ae98c68c0e5c6

dezgeg replied:
Oops, yes, that's it. Also I didn't notice it needs to be a list, looks great now.

nshalman committed:
installer: netboot: fix some missing files

Addresses feedback from @dezgeg.
With help from @bobvanderlinden.

nshalman commented:
Added another commit with the suggested changes from @dezgeg (as fixed up by @bobvanderlinden).
When all of the code is in a good state I can squash things down before this gets merged.

domenkozar commented:
From NixOS release point of view, this needs a section in the manual how it's used. Without it, many people will try and give up, which is a bad demotivator after the excitement about this feature.

nshalman commented:
@domenkozar, it's unclear if you're saying that this will not be merged until the branch includes documentation, but if that is the case please provide more specific instructions as to where that documentation needs to live and how detailed it needs to be.

Below is some sample documentation text. I'm not sure how much detail we should provide as it could devolve into documenting how to set up an entire PXE boot infrastructure which seems out-of-scope.

# Installing NixOS using the "netboot" media
These instructions assume that you have an existing PXE or iPXE infrastructure and simply want to add the NixOS installer as another option.

1. Obtain netboot media consisting of:
 * ```bzImage``` -- the Linux kernel
 * ```initrd``` -- the initrd file
 * ```netboot.ipxe``` -- an example ipxe script demonstrating the appropriate kernel command line arguments for this image
2. Place the ```bzImage```, ```initrd```, and optionally, the ```netboot.ipxe``` script on your HTTP/TFTP server
3. Rename and/or customize the ```netboot.ipxe``` script with any required paths before the file names
4. If you are not using the included ```netboot.ipxe``` file, make sure that when booted the kernel will get the appropriate kernel command line arguments that are included in it.
5. Boot over the network
6. Follow the NixOS installation guide as though you had booted from a CD.

nshalman committed:
remove extraneous console setting from ipxe script

domenkozar commented:
@nshalman thanks! I think documentation should be a requirement for such changes, but it's currently not. I still think requirement won't help much, but we should rather talk and agree that it's necessary for a healthy project to succeed (what we all strive for).

The best place for this would be in a new section at https://github.com/NixOS/nixpkgs/tree/master/nixos/doc/manual/installation

bobvanderlinden commented:
There are 2 parts for the documentation: iPXE installer and the iPXE module. The documentation for the installer should cover all steps to build and boot the iPXE file:

--

To build the necessary files use:
```
nix-build -A netboot <nixos/release.nix>
```
This will result in a directory with `bzImage`, `initrd` and `netboot.ipxe`.

To make the files available over HTTP we use:
```
darkhttpd result/
```
In this example we assume `darkhttpd` is listening on `http://myhost:8080/`.

Now boot a system using iPXE, press `Ctrl+b` upon startup to enter the iPXE command line. Next enter the following commands:

```
dhcp
chain --autofree http://myhost:8080/netboot.ipxe
```
For more information see [the iPXE documentation](http://ipxe.org/cmdline).

--

Though `darkhttpd` might not be as well-known to be simple for most people. It is also somewhat the manual way of doing things.

Ideally people shouldn't need to host the installer themselves and Hydra would host the 3 files. People just need to use iPXE to boot the NixOS installer that way. Or even simpler, use netboot.xyz.

As for the module, the NixOS manual doesn't document each module separately, but it does refer to https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix, so I guess `netboot.nix` should be added there.

cleverca22 commented:
https://gist.github.com/cleverca22/6077ceaeaaa93470ff1810499c534dc7

in the case where the end-user controls the dhcp server, you also have this config as an option

this tells the firmware on the computer to boot iPXE from a tftp server, and then tells iPXE server to load a given script over http

fpletz commented:
Tested on 2 hardware devices, one SuperMicro server with an Intel NIC and a Thinkpad. Works like a charm like any other iPXE payload for us. Thank you very much! :+1:

nshalman commented:
I'm working on documenting this, but I must say that docbook isn't the most pleasant authoring experience. I realize the strange loop in this question, but is there any documentation on how to write NixOS documentation?

domenkozar commented:
I feel your pain. Feel free to write it in markdown and I'll convert it.

nshalman commented:
Here's what I have so far as a combination of my own original version and @bobvanderlinden's version.
Feedback on it is welcome. Once we're happy with it, I'm happy to get a converted version from @domenkozar and add it to the branch myself, or we can merge documentation separately.

domenkozar commented:
@nshalman looks good. I'd open an issue for [copying the iPXE installer to releases](http://nixos.org/releases/nixos/unstable/nixos-16.09pre79453.32b7b00/) and link it in the last "maybe" section of documentation.

Also, would be nice to get an OK from @edolstra and then I'll convert the documentation into docbook and merge this.

domenkozar commented on nixos/modules/installer/netboot/netboot.nix:
```diff
@@ -0,0 +1,93 @@
+
+    boot.initrd.availableKernelModules = [ "squashfs" ];
+
+    boot.blacklistedKernelModules = [ "nouveau" ];
```
Would be nice to have a comment here for reasoning.

nshalman replied:
I think that code came along for the ride from
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/cd-dvd/iso-image.nix#L287

Which is blamed on https://github.com/NixOS/nixpkgs/commit/6948c3ab80797770bc584e521b5c5a8ca02b495e

Which reveals that @edolstra "observed a hang at the moment the nouveau driver gets loaded on a GTX 970."

Not sure what to do with that...

bobvanderlinden replied:
See @edolstra comment: see https://github.com/NixOS/nixpkgs/commit/6948c3ab80797770bc584e521b5c5a8ca02b495e#commitcomment-12369327

Maybe create a separate PR for removing the `boot.blacklistedKernelModules` altogether? (probably could do that after this PR too)

nshalman replied:
Good find, @bobvanderlinden. I can remove that line, then.

domenkozar commented:
@nshalman thank you for this - we're very close :-)

nshalman commented:
@domenkozar would you mind opening the issue for copying the iPXE installer to releases?
I'm not totally clear on what you want in there.

nshalman committed:
installer: netboot: don't blacklist nouveau

nshalman commented:
Just a reminder that I'm happy to squash this branch back down to 2 commits before merging.

domenkozar commented:
@nshalman sure: https://github.com/NixOS/nixos-channel-scripts/issues/6

domenkozar commented:
@nshalman can you squash?

domenkozar commented:
Pushed the changes in 83c0aca062369aa43d5e692f22ff22e8d2e787a1 and updated manual+changelog in 90d7041ce4d270bbcf0efefd1d1e42c097c1335d. Thank you!

edolstra commented:
This gives an eval error:
```
$ nix-build nixos/release.nix -A netboot --show-trace
error: value is a string while a set was expected, at /home/eelco/Dev/nixpkgs/nixos/release.nix:114:16
```
I think there have been some changes to `symlinkJoin`.

nshalman commented:
I'm seeing that error too. Once I figure out what's going on I'll open a follow up to fix that.

nshalman commented:
Looks like @domenkozar already fixed it in 640ec5f6491ad9b6c37d12db0450745d9c7c59c7

Sorry I wasn't around on Friday for the squashing, but I'm glad this work has finally landed.

Thank you again to everyone who helped me with this work.

bobvanderlinden commented:
Awesome work! Is the output now also available through hydra? It didn't seem like it when I tried here: http://hydra.nixos.org/build/34862738#tabs-summary

Whatever the case, the next step for netbooting over internet is https://github.com/NixOS/nixos-channel-scripts/issues/6: making the files available on http://nixos.org/releases/.
